File: aipod-node/src/agent/sdk-reference.ts
/** Versioned SDK reference injected by role; examples are compiled and run in tests. */
import type { StageName } from "./types.js";

export const SDK_EXAMPLES = {
  model: `export interface ExampleValue { value: number }
`,
  provider: `import type { ConfigStore } from "aipod-node";

export class ConfiguredProvider {
  constructor(private readonly deps: { ConfigStore: ConfigStore }) {}
  add(value: number): number {
    return value + Number(this.deps.ConfigStore.get("example.increment", 2));
  }
}
`,
  repository: `import type { ModelRepository } from "aipod-node";

export async function repositoryExample(repo: ModelRepository) {
  const saved = await repo.save("examples", { id: "one", value: 3 });
  const found = await repo.get("examples", saved.id);
  const matches = await repo.find("examples", { value: 3 });
  const deleted = await repo.delete("examples", saved.id);
  return { found, matches, deleted };
}
`,
  contracts: `import type { Contract } from "aipod-node";
export const inputs = { value: { type: "integer", required: false } } as const satisfies Contract;
export const outputs = { answer: { type: "integer" } } as const satisfies Contract;
`,
  service: `import type { PipelineContext } from "aipod-node";
import type { ConfiguredProvider } from "../../providers/public/example.js";

export class ExampleService {
  constructor(private readonly deps: { ConfiguredProvider: ConfiguredProvider }) {}
  execute(context: PipelineContext) {
    const ctx = context.typed(
      { value: { type: "integer", required: false } },
      { answer: { type: "integer" } },
    );
    return ctx.output({ answer: this.deps.ConfiguredProvider.add(ctx.get("value") ?? 1) });
  }
}
`,
  pipeline: `import { service, type Container } from "aipod-node";

export function createPipeline(container: Container) {
  return service(container, "ExampleService");
}
export function createRoute(container: Container) {
  return { name: "example", pipeline: createPipeline(container) };
}
`,
  runner: `import { loadRunner } from "aipod-node";

export async function runExample(projectRoot: string) {
  const runner = await loadRunner(projectRoot);
  const { result, context } = await runner.run("example", { value: 3 });
  if (result.status === "failure") throw new Error(result.error.message);
  return { output: result.output, context };
}
`,
  adapter: `import type { InterfaceAdapter, PipelineRunner } from "aipod-node";

export class ExampleAdapter implements InterfaceAdapter {
  constructor(private readonly runner: PipelineRunner) {}
  requiredRoutes(): string[] { return ["example"]; }
  start(payload: Record<string, unknown> = {}) {
    return this.runner.run("example", payload);
  }
}
`,
  interface_check: `import { loadInterface, smokeInterface } from "aipod-node";

export async function checkInterface(projectRoot: string) {
  const report = await smokeInterface(projectRoot, "Example");
  if (report.status !== "passed") throw new Error(JSON.stringify(report));
  const { adapter, context } = await loadInterface(projectRoot, "Example");
  return { adapter, context };
}
`,
} as const;

const example = (name: keyof typeof SDK_EXAMPLES) => "\n```typescript\n" + SDK_EXAMPLES[name] + "```\n";

export const SDK_SECTIONS = {
  models: `MODEL API
Node Models are ordinary TypeScript interfaces/types/classes; there is no Python-style Model
base class. Interfaces are erased at runtime: do not instantiate or inject them. Register their
model IDs/files for the project ledger, and import their types from the registered source.
Use explicit runtime/domain validation where TypeScript's compile-time types are insufficient.
` + example("model"),
  dependencies: `CONFIGURATION, DEPENDENCIES AND STORAGE
Source imports use "aipod-node". Provider/Service constructors receive ONE object keyed by
exact dependency IDs, e.g. deps.ConfigStore or deps.ModelRepository, not positional providers.
The Container resolves Providers/Services as singletons; Models are not injectable.
ConfigStore(projectRoot, file?) requires await config.load() when created manually;
loadContainer/loadRunner already load it. config.get(path, fallback?) -> value/fallback;
await config.reload() reloads configuration. Node config is config.json/config.toml.
ModelRepository(projectRoot, file=".aipod/data.json") is the built-in JSON store.
await repo.save(collection, value, idField="id") -> saved record; id must be string/number.
await repo.get(collection, id) -> record | undefined; await repo.list(collection) -> records.
await repo.find(collection, filters) -> matching records; await repo.delete(collection, id) -> boolean.
Generic record values must satisfy Record<string, unknown>; spread a typed interface into a
plain record when necessary. Use AIPOD_DATA_DIR for temporary data while checking generation.
` + example("repository"),
  providers: "PROVIDER IMPLEMENTATION\nPlace the class in impl/ and expose a thin named re-export in public/.\n" + example("provider"),
  context: `PIPELINE CONTEXT
new PipelineContext(params={}, {data?, branchId?}={}).
context.get(key, fallback?) -> data first, then params, then fallback; set(key, value) -> void.
context.summary() -> {params, data, steps}; context.steps stores execution traces.
Services use context.typed(literalInputs, literalOutputs) -> ContractContext with typed get/set/output.
ctx.get(name) checks a declared input; ctx.set(name, value) writes a checked declared output.
ctx.output(mapping) validates and returns the complete declared output mapping; it does not
itself publish it to the underlying context. Return it from execute(), so the runtime publishes it.
Keep schemas literal in context.typed(...) for field inference; do not cast contracts to any.
`,
  contracts: `REGISTRATION CONTRACTS
Contract = Record<string, ContractField>. Each field is an object with type/required/properties/items.
Types: string, number, integer, boolean, object, array, any. required defaults to true;
required:false permits an absent key, not null for a string/number field. Node contracts have
no automatic default injection; use ?? or domain validation in the Service for optional inputs.
Nested object properties are another Contract; array items are a ContractField.
Do not use Python type strings or prose such as "int 1..60, optional" as a Node field spec.
Specify business ranges explicitly in your domain code. Match finish metadata to the literal
schemas in execute(). Outputs missing on a failure path must be optional or consistently supplied.
validateContract(data, contract={}, prefix="$") -> string[]; [] means valid.
Register Provider/Service public files via finish; the controller maintains aipod.json.
` + example("contracts"),
  services: "SERVICE IMPLEMENTATION\nThe dependency is an existing public Provider; no Service-to-Service calls.\n" + example("service"),
  pipelines: `COMPOSITION API
Import service, sequence, parallel, repeat and type Container from "aipod-node".
service(container, "ServiceId") -> ServiceRef; sequence(...nodes) -> sequential ExecutionNode.
await node.execute(context) -> Result. ServiceRef.retry(retries=3, delayMs=0) and
fallback(otherRef) return configured references. Compose Services only in this layer.
Export createPipeline(container) and createRoute(container) in a pipeline source file.
Before registration, test the factory with a real container. Submit name/file/services/execution
and public inputs in finish.routes. loadRunner builds registered routes from aipod.json's
services/execution metadata; keep that metadata consistent with the generated factory.
` + example("pipeline"),
  runner: `REGISTERED ROUTE EXECUTION
await loadContainer(projectRoot) -> Container; container.resolve(id) -> Provider/Service instance.
await loadRunner(projectRoot) -> PipelineRunner (compiles project sources and loads registered beans).
Reuse a loaded runner in an HTTP server instead of recompiling for every request.
runner.routeNames() -> string[]; await runner.run(routeName, params={}) -> {result, context}.
result is a discriminated Result: success has status="success", output, effects;
failure has status="failure", error={code,message,retryable,details}, effects.
Service errors normally become failure Results; unknown routes and loading/compilation can throw.
An application result may have its own ok/error inside result.output: inspect that too.
Do not treat the outer {result, context} as the output itself or as Python's raw dict result.
typeCheckProject(projectRoot) -> Promise<SemanticDiagnostic[]>; [] means no TS diagnostics.
` + example("runner"),
  interfaces: `INTERFACE SDK — Node signatures and return shapes
InterfaceAdapter is a TypeScript interface: implement it, do not extend it as a runtime class.
requiredRoutes(): string[]; start(payload?: Record<string, unknown>): unknown | Promise<unknown>.
The loader passes a PipelineRunner to the Adapter constructor, not a dependencies dictionary.
For manifest name "Example", export class ExampleAdapter from the registered file.
await loadInterface(projectRoot, name) -> {adapter, context}; it compiles/loads the project.
new InterfaceContext(projectRoot); await context.routeNames() -> string[];
await context.runRoute(route, params={}) -> {result, context}; await context.runner() -> PipelineRunner.
await adapter.start(payload) takes ONLY the payload: do not pass a Python-style InterfaceContext.
start may keep an HTTP/GUI/worker running. Test such entries in a child process, probe them,
then terminate and wait in finally; do not await a long-running start before sending requests.
smokeInterface(projectRoot, name) -> Promise<{status, requiredRoutes, missingRoutes}> checks
route presence only, not functional behavior. There is no universal Node adapter.stop() method.
Interface metadata goes in finish.interfaces: name/file/route/kind/artifacts, plus any intended
lifecycle/verify/permissions. Preserve metadata on updates. lifecycle install/uninstall commands
are argv arrays with {node}/{projectRoot} placeholders, not joined shell strings.
verify entries are {name, command: string[], required: boolean, timeoutMs?} objects.
await verifyInterface(projectRoot, name) -> {status, checks}; an empty checks array proves no behavior.
` + example("adapter") + example("interface_check"),
  verification: `ENTRY AND ACCEPTANCE CHECKS
During Agent checks: node "$AIPOD_NODE_CLI" interface smoke Example --project-root /PROJECT
node "$AIPOD_NODE_CLI" interface run Example --payload '{"value":3}' --project-root /PROJECT
The smoke/verify CLI prints a status object: inspect status/checks, not only its exit code.
For user projects, the equivalent installed binary is aipod-node.
runVerificationCommand(projectRoot, argv, timeoutMs=120000) -> Promise<{status, exitCode,
stdout, stderr, command, durationMs}>. argv is executed without a shell; | is not a pipeline there.
Run actual positive/negative behavior checks and rerun the failed check after its owner repairs it.
Do not mask test exit codes with | tail or || true; the framework already bounds captured output.
A successful import, help command or route-presence smoke is not full acceptance. Report the
checks actually performed and remaining gaps, rather than claiming all requirements passed.
`,
} as const;

export const SDK_ROLE_SECTIONS = {
  models: ["models"],
  providers: ["dependencies", "contracts", "providers"],
  services: ["context", "dependencies", "contracts", "services"],
  pipelines: ["context", "contracts", "pipelines", "runner"],
  interfaces: ["contracts", "runner", "interfaces", "verification"],
  pod: ["contracts", "runner", "interfaces", "verification"],
} as const;

export function sdkReference(role: StageName | "pod"): string {
  if (!Object.hasOwn(SDK_ROLE_SECTIONS, role)) throw new Error(`Unknown SDK reference role: ${role}`);
  return `AIPod Node SDK reference bundled with this framework — role: ${role}.
Use these signatures and examples directly; these SDK snippets are not response instructions.
Example names/paths are placeholders: use the registered names and dependencies in project context.
Preserve AIPOD_NODE_CLI/AIPOD_NODE_MODULE/AIPOD_BUILD_DIR; for one-off ESM checks use
node --input-type=module and await import(process.env.AIPOD_NODE_MODULE).
This reference remains in the system prompt on every turn. Inspect SDK source only for a
specific missing detail or observed discrepancy, not to repeatedly rediscover these APIs.
Other chapters are available through sdkReference(role), exported by aipod-node.
\n` + SDK_ROLE_SECTIONS[role].map(name => SDK_SECTIONS[name].trim()).join("\n\n");
}

File: ai_pod_cli/sdk_reference.py
"""Versioned SDK reference injected by role; examples are executed in tests."""

from textwrap import dedent


SDK_EXAMPLES = {
    "model": '''from ai_pod_cli import Model

class ExampleValue(Model):
    value: int
''',
    "provider": '''from injector import inject
from ai_pod_cli.config_store import ConfigStore

class ConfiguredProvider:
    @inject
    def __init__(self, config: ConfigStore):
        self.increment = config.get("example.increment", 2)

    def add(self, value: int) -> int:
        return value + self.increment
''',
    "repository": '''def repository_example(repo, Record):
    # Record is an existing registered table Model with id and value fields.
    saved = repo.save(Record(value=3))
    found = repo.get(Record, saved.id)
    matches = repo.find(Record, value=3)
    repo.delete(saved)
    return found, matches
''',
    "contracts": '''INPUTS = {"value": {"type": "int", "required": False}}
OUTPUTS = {"answer": {"type": "int"}}
''',
    "service": '''from injector import inject
from ai_pod_cli import PipelineContext
from modules.providers.public.example import ConfiguredProvider

class ExampleService:
    @inject
    def __init__(self, provider: ConfiguredProvider):
        self.provider = provider

    def execute(self, ctx: PipelineContext):
        result = {"answer": self.provider.add(ctx.get("value", 1))}
        ctx.set("answer", result["answer"])
        return result
''',
    "pipeline": '''from ai_pod_cli.config import load_beans
from ai_pod_cli.container import Pod, build_container
from modules.services.public.example import ExampleService

def run(ctx):
    S = Pod(build_container(load_beans()))
    return S(ExampleService).execute_all(ctx)
''',
    "runner": '''from ai_pod_cli.runner import PipelineRunner

def run_example():
    runner = PipelineRunner()
    result = runner.run("example", {"value": 3})
    return result
''',
    "adapter": '''from ai_pod_cli.interface import InterfaceAdapter

class ExampleAdapter(InterfaceAdapter):
    def required_routes(self):
        return ["example"]

    def smoke_payloads(self):
        return {"example": {"value": 3}}

    def start(self, context, payload=None):
        # A short-lived entry; HTTP/GUI adapters may instead block in start().
        return context.run_route("example", payload or {})
''',
    "interface_check": '''from ai_pod_cli.interface import load_manifest, load_adapter, create_context

def check_interface(project_root="."):
    manifest_path, manifest = load_manifest("example", project_root)
    adapter = load_adapter(manifest, project_root)
    context = create_context(manifest, project_root)
    report = adapter.smoke(context)
    assert report["status"] == "passed", report
    return manifest_path, adapter, context
''',
}


def _example(name):
    return "\n```python\n" + SDK_EXAMPLES[name] + "```\n"


SECTIONS = {
    "models": dedent('''
        MODEL API
        from ai_pod_cli import Model; from sqlmodel import Field
        Model inherits SQLModel/Pydantic. Define value models without table=True.
        Persistent models use class Record(Model, table=True) and a primary key such as
        id: int | None = Field(default=None, primary_key=True). Import existing Models by
        registered class_path; they are data, not injected dependencies.
        ModelClass.model_validate(value) -> instance; raises Pydantic ValidationError.
        ModelClass.validate(value, path="$" ) -> list[str]; [] means valid, not an instance.
        instance.to_dict() -> dict. ModelClass.sample() -> synthetic required-field values;
        samples do not establish domain validity. Add strictness/range validators as needed.
    ''') + _example("model"),
    "dependencies": dedent('''
        CONFIGURATION, DEPENDENCIES AND STORAGE
        from ai_pod_cli.config_store import ConfigStore
        from ai_pod_cli.repository import ModelRepository
        Python constructors use @inject from injector and actual Provider type annotations.
        The container binds Providers/Services as singletons. Register dependency IDs in
        finish.components; Python injects by type, not a Node-style dependencies dictionary.
        ConfigStore(config_path="config.toml") loads TOML immediately.
        config.get("section.key", default=None) -> value/default; config.reload() reloads it.
        ModelRepository(config_store: ConfigStore) uses AIPOD_DATABASE_URL or database.url.
        repo.save(instance) -> saved Model (generated id refreshed).
        repo.get(ModelClass, id) -> Model | None; repo.list(ModelClass) -> list[Model].
        repo.find(ModelClass, filters=None, **filter_values) -> list[Model].
        repo.delete(instance) -> None; repo.close() releases connections.
        Save/get/list/find initialize tables from modules.models. Reuse the built-in repository
        for persistence; use the Agent's temporary database for checks.
    ''') + _example("repository"),
    "providers": "PROVIDER IMPLEMENTATION\nPlace the implementation in impl/ and re-export it by name from public/.\n" + _example("provider"),
    "context": dedent('''
        PIPELINE CONTEXT
        from ai_pod_cli import PipelineContext
        PipelineContext(params=None, *, data=None, branch_id=None).
        ctx.get(key, default=None) reads data first, then params. ctx.set(key, value) -> None.
        ctx.summary() -> {params, data, steps}; ctx.steps contains execution traces.
        Python PipelineContext has no output() or typed() method.
        Return the complete declared output mapping from Service.execute(), even if those
        values were also written with ctx.set: execute_all() returns the Service's raw result.
    '''),
    "contracts": dedent('''
        REGISTRATION CONTRACTS
        inputs/outputs are mappings of field names to schemas. Use plain types (int, float,
        str, bool, dict, list, any) or structured schemas. required=False permits an absent key;
        it does not permit None for an int/str field. Nullable values can use an anyOf schema
        such as {"anyOf": [{"type": "str"}, {"type": "null"}]}.
        Do not put prose such as "int 1..60, optional" in a type expression; put explanations
        in description and implement domain bounds explicitly. A string type implies required.
        Do not assume contract defaults populate ctx: use ctx.get(name, fallback) for defaults.
        Outputs absent on failure paths must be optional, or returned on every path.
        validate_contract_data(data, fields, prefix="$") -> list[str] in ai_pod_cli.contracts.
        Submit public Provider/Service class_path and matching schemas through finish;
        registrations are written by the controller, not by an SDK call in generated code.
    ''') + _example("contracts"),
    "services": "SERVICE IMPLEMENTATION\nThis example uses the contracts above and an existing public Provider.\n" + _example("service"),
    "pipelines": dedent('''
        COMPOSITION API
        from ai_pod_cli.config import load_beans; from ai_pod_cli.container import Pod, build_container
        load_beans() -> {beans: [...]}; build_container(config) -> injector.Injector.
        S = Pod(container); S(ServiceClass) returns a composition reference. Pod is not a
        context manager. Use classes imported from their registered public paths.
        S(First).execute_all(ctx) -> raw result; (S(First) | S(Second)).execute_all(ctx) ->
        last result, stopping at Failure. For async Services use await execute_all_async(ctx).
        Define run(ctx) in each pipeline file. Before a NEW route is registered, test run(ctx)
        directly. Submit name/file/inputs in finish.pipelines; the controller writes routes.toml.
    ''') + _example("pipeline"),
    "runner": dedent('''
        REGISTERED ROUTE EXECUTION
        from ai_pod_cli.runner import PipelineRunner
        PipelineRunner(routes_path="routes.toml"); runner.route_names() -> list[str].
        runner.run(route_name, params=None) -> pipeline result dict, NOT a tuple or Node Result.
        runner.run_with_context(route_name, params=None) -> (result, PipelineContext).
        await runner.run_async(...) and run_with_context_async(...) support async pipelines.
        Unknown routes raise KeyError; missing pipeline files raise FileNotFoundError;
        Service/pipeline exceptions can propagate. SDK Success/Failure results are serialized
        with status/output or status/error; ordinary dict results keep their application shape.
        Run from the project root (route file paths are relative), or use InterfaceContext.run_route.
        Construct a new runner after changing route registrations. Match route names and inputs
        to the current registry, and inspect the actual application result when checking success.
    ''') + _example("runner"),
    "interfaces": dedent('''
        INTERFACE SDK — Python signatures and return shapes
        from ai_pod_cli.interface import load_manifest, load_adapter, create_context, InterfaceAdapter
        load_manifest(target, project_root=".") -> (Path, manifest_dict). Unpack both values;
        do not call .get() on the tuple. target may be an Interface name or manifest path.
        load_adapter(manifest_dict, project_root=".") -> instantiated InterfaceAdapter.
        create_context(manifest_dict, project_root=".") -> InterfaceContext.
        context.route_names() -> list[str]; context.route_contract(route) -> contract dict.
        context.run_route(route, params=None) -> pipeline result; it activates the project root.
        context.activated() is a context manager; context.emit(event, payload=None) records events.
        InterfaceAdapter is a base class. Override required_routes(), smoke_payloads() and
        start(self, context, payload=None). Base smoke(context) -> {status: passed|failed,
        missing_routes, missing_smoke_payloads, contract_errors, ...}; it checks route declarations
        and sample input contracts, not the actual HTTP/GUI behavior.
        start() is adapter-defined and may block in a server/event loop. Do not call a blocking
        start() and then try to send HTTP requests in the same thread. Test long-lived entries
        in a separate process, send requests, then terminate/wait for that process in finally.
        Base stop(context) does nothing; a generated adapter must implement actual shutdown
        if it advertises that capability. Keep project activation for the entry lifetime or
        run a child with cwd=context.project_root.
        Manifest: interfaces/<name>/interface.json, with name/kind/adapter/artifacts.
        adapter = {path: "interfaces/example/adapter.py", class_name: "ExampleAdapter"}.
        When returning finish.interfaces include all intended lifecycle/verify/support fields:
        the controller writes the submitted manifest. Lifecycle commands are argv lists,
        e.g. ["{python}", "app.py", "--port", "8770"], not one joined shell string.
        The optional legacy helper verify_adapter_candidate(project_root, manifest, sources,
        timeout=30) -> list[str] requires sources={relative_path: complete_source} and copies
        a candidate project. It is not required for checking an already delivered Interface.
    ''') + _example("adapter") + _example("interface_check"),
    "verification": dedent('''
        ENTRY AND ACCEPTANCE CHECKS
        python -m ai_pod_cli interface --project-root /PROJECT smoke example
        python -m ai_pod_cli interface --project-root /PROJECT --payload '{"value":3}' run example
        The SDK smoke CLI prints a status object: inspect status, not just the process exit code.
        Run actual behavior checks for the original requirement, including negative cases.
        Rerun a failed check after the owner repairs it. Keep test commands' failing exit codes;
        do not hide them with | tail or || true (the framework already bounds captured output).
        A successful import, help command or smoke declaration check is not full acceptance.
        Report which checks ran and what remains unverified. Never claim uncovered behavior passed.
    '''),
}

ROLE_SECTIONS = {
    "models": ("models",),
    "providers": ("dependencies", "contracts", "providers"),
    "services": ("context", "dependencies", "contracts", "services"),
    "pipelines": ("context", "contracts", "pipelines", "runner"),
    "interfaces": ("contracts", "runner", "interfaces", "verification"),
    "pod": ("contracts", "runner", "interfaces", "verification"),
}


def sdk_reference(role: str) -> str:
    """Return only the SDK chapters relevant to this Agent's ownership role."""
    if role not in ROLE_SECTIONS:
        raise ValueError(f"Unknown SDK reference role: {role}")
    header = f'''AIPod Python SDK reference bundled with this framework — role: {role}.
Use these signatures and examples directly; they describe SDK calls, not response instructions.
Example names/paths are placeholders: use the actual registered names from project context.
Use the task's Python interpreter (AIPOD_PYTHON) and preserve its supplied PYTHONPATH.
This reference stays in the system prompt on every turn. Consult SDK source only for a specific
missing detail or observed discrepancy; do not repeatedly dump modules to rediscover these APIs.
Other chapters are available from ai_pod_cli.sdk_reference.sdk_reference(role).
'''
    return header + "\n" + "\n".join(SECTIONS[name].strip() for name in ROLE_SECTIONS[role])
